File: mcp/scripts/check-sdk-live.py
"""Opt-in SDK integration test on a user-authorized test PCB; removes its DOCUMENT-layer marker."""
import asyncio
import json
import struct
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / 'dist' / 'lib' / 'python'))
from easyeda_copilot import connect, Blob, TypedArray
from easyeda_copilot.constants import EPCB_LayerId


async def main():
    if len(sys.argv) < 2:
        raise SystemExit('Usage: python scripts/check-sdk-live.py <test-pcb-document-uuid> [instance-id]')
    report = {'documentUuid': sys.argv[1], 'checkpoints': {}}
    async with await connect(document_uuid=sys.argv[1], instance_id=sys.argv[2] if len(sys.argv) > 2 else None) as session:
        eda = session.eda
        report['checkpoints']['baseline'] = session.last_checkpoint
        components = await eda.pcb_PrimitiveComponent.getAll()
        coordinates = await asyncio.gather(*(getter() for c in components for getter in [c.getState_X, c.getState_Y]))
        raw = struct.pack('<' + 'd' * len(coordinates), *coordinates)
        echoed = await session.eval('return inputs.positions', {'positions': TypedArray(raw, 'Float64Array')})
        assert echoed.type == 'Float64Array' and echoed.data == raw
        before = await eda.pcb_PrimitiveLine.getAllPrimitiveId()
        marker = None
        try:
            marker = await eda.pcb_PrimitiveLine.create('', EPCB_LayerId.DOCUMENT, 0, 0, 10, 0, 1, False)
            assert marker
            report['checkpoints']['create'] = session.last_checkpoint
            marker_id = await marker.getState_PrimitiveId()
            modified = await eda.pcb_PrimitiveLine.modify(marker, {'endX': 30})
            assert modified
            report['checkpoints']['modify'] = session.last_checkpoint
            assert await (await eda.pcb_PrimitiveLine.get(marker_id)).getState_EndX() == 30
        finally:
            if marker:
                assert await eda.pcb_PrimitiveLine.delete(marker)
                report['checkpoints']['delete'] = session.last_checkpoint
        assert sorted(await eda.pcb_PrimitiveLine.getAllPrimitiveId()) == sorted(before)
        preview = await eda.dmt_EditorControl.getCurrentRenderedAreaImage()
        assert isinstance(preview, Blob) and preview.type == 'image/png' and preview.size > 0
        assert await session.eval('return inputs.image', {'image': preview}) == preview
        report.update(components=len(components), originalLineIdsPreserved=True, typedCoordinatesRoundTrip=True,
                      previewRoundTrip=True, previewBytes=preview.size)
    folder = ROOT / '.test-data' / 'sdk-live'
    folder.mkdir(parents=True, exist_ok=True)
    (folder / 'python-report.json').write_text(json.dumps(report, indent=2), encoding='utf-8')
    print(json.dumps(report))


asyncio.run(main())

File: mcp/scripts/check-sdk-live.mjs
// Opt-in integration check against an open TEST PCB. Creates and removes one DOCUMENT-layer line.
import assert from 'node:assert/strict';
import { mkdir, writeFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { connect, EPCB_LayerId } from '../dist/lib/node/index.mjs';

const documentUuid = process.argv[2];
if (!documentUuid) throw Error('Usage: node scripts/check-sdk-live.mjs <test-pcb-document-uuid> [instance-id]');
const start = Date.now();
const session = await connect({ documentUuid, instanceId: process.argv[3] });
const report = { documentUuid, instanceId: session.instanceId, checkpoints: {} };
let line;
try {
    report.checkpoints.baseline = session.lastCheckpoint;
    const components = await session.eda.pcb_PrimitiveComponent.getAll();
    const designators = await Promise.all(components.map(c => c.getState_Designator()));
    report.components = components.length;
    report.designators = designators.slice(0, 8);
    const before = await session.eda.pcb_PrimitiveLine.getAllPrimitiveId();
    line = await session.eda.pcb_PrimitiveLine.create('', EPCB_LayerId.DOCUMENT, 0, 0, 10, 0, 1, false);
    report.checkpoints.create = session.lastCheckpoint;
    assert(line, 'Native create returned undefined');
    report.createdId = await line.getState_PrimitiveId();
    const modified = await session.eda.pcb_PrimitiveLine.modify(line, { endX: 20 });
    report.checkpoints.modify = session.lastCheckpoint;
    assert(modified);
    const reread = await session.eda.pcb_PrimitiveLine.get(report.createdId);
    assert.equal(await reread.getState_EndX(), 20);
    assert.equal(await session.eda.pcb_PrimitiveLine.delete(line), true);
    line = undefined;
    report.checkpoints.delete = session.lastCheckpoint;
    const after = await session.eda.pcb_PrimitiveLine.getAllPrimitiveId();
    assert.deepEqual([...after].sort(), [...before].sort());
    report.originalLineIdsPreserved = true;
    const preview = await session.eda.dmt_EditorControl.getCurrentRenderedAreaImage();
    assert(preview instanceof Blob, 'Preview must be a local native Blob');
    assert(preview.size > 0);
    const folder = fileURLToPath(new URL('../.test-data/sdk-live/', import.meta.url));
    await mkdir(folder, { recursive: true });
    report.preview = { type: preview.type, bytes: preview.size, path: resolve(folder, 'preview.png') };
    await writeFile(report.preview.path, Buffer.from(await preview.arrayBuffer()));
    report.durationMs = Date.now() - start;
    await writeFile(resolve(folder, 'report.json'), JSON.stringify(report, null, 2));
    console.log(JSON.stringify(report, null, 2));
} finally {
    try { if (line) await session.eda.pcb_PrimitiveLine.delete(line); }
    finally { await session.close(); }
}
